File: prettier-plugin-csharpier/src/Printer/Types/WhileStatement.ts
import { PrintMethod } from "../PrintMethod";
import { SyntaxToken, printPathSyntaxToken, SyntaxTreeNode } from "../SyntaxTreeNode";
import { concat, group, hardline, indent, join, softline, line, doubleHardline } from "../Builders";

export interface WhileStatementNode extends SyntaxTreeNode<"WhileStatement"> {
    whileKeyword: SyntaxToken;
    condition: SyntaxTreeNode;
    statement: SyntaxTreeNode;
}

export const printWhileStatement: PrintMethod<WhileStatementNode> = (path, options, print) => {
    return concat([
        printPathSyntaxToken(path, "whileKeyword"),
        " (",
        path.call(print, "condition"),
        ")",
        path.call(print, "statement")
    ]);
};

File: Src/CSharpier/SyntaxPrinter/SyntaxNodePrinters/WhileStatement.cs
namespace CSharpier.SyntaxPrinter.SyntaxNodePrinters;

internal static class WhileStatement
{
    public static Doc Print(WhileStatementSyntax node, FormattingContext context)
    {
        var result = Doc.Concat(
            ExtraNewLines.Print(node),
            Doc.Group(
                Token.Print(node.WhileKeyword, context),
                " ",
                Token.Print(node.OpenParenToken, context),
                Doc.Group(
                    Doc.Indent(Doc.SoftLine, Node.Print(node.Condition, context)),
                    Doc.SoftLine
                ),
                Token.Print(node.CloseParenToken, context)
            ),
            node.Statement switch
            {
                WhileStatementSyntax => Doc.Group(
                    Doc.HardLine,
                    Node.Print(node.Statement, context)
                ),
                _ => OptionalBraces.Print(node.Statement, context),
            }
        );

        return result;
    }
}
